### PR TITLE
[VFX] Add Graphics Buffer support

### PR DESCRIPTION
**Pull request by PaulDemeulenaere**
### Purpose of this PR 
Adding Graphics Bufffer support defined by this [technical design document](https://docs.google.com/document/d/1NOfMAPF7yBt-LQDoPP2UhP_kgUz6mhNgIeO0iwLndos/edit).

> Add Graphics Buffer in VFX Graph as a first class citizen. Graphics Buffer would be like any other exposed type (except not serialized). This will allow more flexibility and simplicity for users to pass and interpret their own data. At the moment they have to rely on textures for that.

Global approach ⬇️ (the generated code isn't up to date in this screenshot)
image

- Add two operators :
_sampling_graphics_buffer _buffer_count
_⚠️ TODO : document newly added operator._

- Expose VFXType with a new optionnal attribute : `Usage.GraphicsBuffer` & moved VFXType to runtime, sample of usage :
```[VFXType]
public struct Rectangle
{
	public Vector2 size;
	public Vector3 color;
}

[VFXType(VFXTypeAttribute.Usage.GraphicsBuffer)]
struct CustomData
{
	public Rectangle rectangle;
	public Vector3 position;
}
```
_⚠️ TODO : document newly exposed API._

---
### Testing status
Adding a new graphic test which is displaying a set of rectangle, see this code : https://github.cds.internal.unity3d.com/unity/vfx-graphics/pull/236/files#diff-6e84b835af122c8437c16eb04557d42aR178
_graphics_test_buffer

Adding several editor test :
- CreateComponent_And_BindGraphicsBuffer : Check bindings & the value of GetNativeBufferPtr consistency
- CreateComponent_And_BindGraphicsBuffer_And_(Reinit) : Check the binding buffer is kept while calling Reinit
- CreateComponent_And_BindGraphicsBuffer_And_(DisableAndRenable) : Check the binding buffer is kept while disabling/renabling visualEffect
- CreateComponent_And_BindGraphicsBuffer_And_(ChangeVisualEffectAsset) : Switch to another visualEffectAsset (which also have the same buffer exposed) [(fixed by this C++ change)](https://ono.unity3d.com/unity/unity/changeset/ffb4a550d59e9fd4be81c5284de751ca9aa94e33)
- CreateComponent_And_BindGraphicsBuffer_And_(EditSerializedObject) : Modifying the serialized property [(fixed by this C++ change)](https://ono.unity3d.com/unity/unity/changeset/ffb4a550d59e9fd4be81c5284de751ca9aa94e33)
- Slot_Sanitize_When_A_Sub_Slot_Is_Missing_Link : Cover a behavior added in [VFXSlot.cs](https://github.com/Unity-Technologies/Graphics/pull/4109/files#diff-3704d727c796d3b0daba6b4c931351e5681dfdd6552ae1349a94b02033c5db33L478), we can correctly sanitize if VFXType change its structure adding a field, if slot count doesn't correspond, try to copy by field name.

Tested locally some invalid VFXType :

- Private field aren't supported in VFXType if GraphicsBuffer usage
```[VFXType(VFXTypeAttribute.Flags.GraphicsBuffer)]
public struct Type_Using_Private_And_GraphicsBuffer
{
	public int a;
	private int b;
}
```
- Use another struct which isn't a VFXType
```public struct NotAVFXType
{
	public int a;
}

[VFXType]
public struct Type_Using_Not_VFX_Type_SubType
{
	public float a;
	public NotAVFXType b;
}
```
- Unsupported blittable type by the VFX
```[VFXType(VFXTypeAttribute.Flags.GraphicsBuffer)]
public struct Type_Using_Not_Supported_Type_By_VFX
{
	public float a;
	public byte b;
}
```
- Not blittable type
```[VFXType(VFXTypeAttribute.Flags.GraphicsBuffer)]
public struct Type_Not_Blittable
{
	public float a;
	public Texture2D b;
}
```
**Note** : I'm expecting other error case but I tried to cover the most obvious possible incompatibilities.

I've run yamato to check the expected result :
- [VFX_URP private ](https://yamato.cds.internal.unity3d.com/jobs/339-ScriptableRenderPipelinePrivate/tree/vfx%252Ffeature%252Fgraphics-buffer-support/.yamato%252Fall-vfx_urp.yml%2523Nightly_VFX_URP_CUSTOM-REVISION/6154171/job)
- [VFX_URP public](https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/vfx%252Ffeature%252Fgraphics-buffer-support/.yamato%252Fall-vfx_urp.yml%2523PR_VFX_URP_CUSTOM-REVISION/6154215/job)
- [VFX_HDRP private](https://yamato.cds.internal.unity3d.com/jobs/339-ScriptableRenderPipelinePrivate/tree/vfx%252Ffeature%252Fgraphics-buffer-support/.yamato%252Fall-vfx_hdrp.yml%2523PR_VFX_HDRP_CUSTOM-REVISION/6154180/job)
- [VFX_HDRP private](https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/vfx%252Ffeature%252Fgraphics-buffer-support/.yamato%252Fall-vfx_hdrp.yml%2523PR_VFX_HDRP_CUSTOM-REVISION/6154222/job)

_⚠️ TODO : upload image references ⚠️_ 

---
### Comments to reviewers
_⚠️ This PR require C++ code : https://ono.unity3d.com/unity/unity/pull-request/123849/_/graphics/vfx/feature/add-graphics-buffer-support_  (not final, still an ownership issue to resolve see this [conversation](https://unity.slack.com/archives/G1BTWN88Z/p1617868925065600))

Known limitation :
- We aren't checking the type of buffer, GraphicsBuffer.Target.Structured is kind of expected view

Known issue (wip) :
- The inspector is displaying the GraphicsBuffer

### CODE REVIEW

**Comment by landonth** [comment]
Suggestion to rename operator node from "Sample Graphics Buffer" to "Load Graphics Buffer." **Sample** implies texture interpolation which is impossible when reading from Graphics Buffer. **Load** is consistent with existing (discrete, integer indexed, no interpolate) behavior and naming of VFX nodes "Load Texture2D," "Load Texture3D," etc

Aside from consistency with VFX Graph's existing nodes and DirectX/Vulkan/OpenGL terminology, Sampling in computer graphics in general is primarily defined as an operation where "continuous functions (signals) are discretized via sampling" or "The process of mapping a continuous function
to a discrete one is called sampling." Whereas here with GraphicsBuffer, we are simply loading already discretized data --not sampling from any continuous space via a continous function or interpolation as the name strongly implies.

[https://doi.org/10.1007/978-3-642-27851-8_167-1](Subr K. (2015) Sampling Problems in Computer Graphics. In: Luo R. (eds) Encyclopedia of Color Science and Technology. Springer, Berlin, Heidelberg.)
[https://www.cs.rpi.edu/~cutler/classes/advancedgraphics/F05/lectures/20_sampling_aliasing.pdf](Sampling, Aliasing, & Mipmaps - CSCI-6962 Advanced Computer Graphics
Cutler)

**Comment by VladNeykov** [comment]
Hi, dropping the WIP [test doc](https://docs.google.com/document/d/1lkQj67dt_kI6PpjS78UoPLlw8wbtvCQ5VabFFeG5r5w/edit).

**Comment by PaulDemeulenaere** [comment]
> Suggestion to rename operator node from "Sample Graphics Buffer" to "Load Graphics Buffer." **Sample** implies texture interpolation or triangle surface interpolation which is not possible when reading custom data from a GraphicsBuffer. **Load** is consistent with existing (discrete, integer indexed, no interpolate) 1:1 data loading behavior and naming of VFX nodes "Load Texture2D," "Load Texture3D," etc. Distinct from nodes with floating point UV coordinate inputs like "Sample Texture2D," etc.
> 
> The only reason I can think of the keep the "Sample" name is if there is planned functionality for software interpolation built-in to the Sample Graphics Buffer node for certain common GraphicsBuffer data layouts. i.e. Triangle, Quad primitives, regular 1D, 2D, and 3D grids. In the same way that the Sample Mesh and Sample Skinned Mesh nodes can optionally sample vertex interpolated data from the triangle mesh surface. Is this the case?
> 
> This is probably well known to you, but I will cite my sources for completeness. Aside from consistency with VFX Graph's existing nodes and DirectX/Vulkan/OpenGL terminology, Sampling in computer graphics is primarily defined as an operation where "continuous functions (signals) are discretized via sampling" or "The process of mapping a continuous function to a discrete one is called sampling." Whereas here with GraphicsBuffer, we are (as far as I can tell) simply loading already discretized data --not sampling from a continuous space via a function or interpolation as the name strongly implies.
> 
> * [Subr K. (2015) Sampling Problems in Computer Graphics. In: Luo R. (eds) Encyclopedia of Color Science and Technology. Springer, Berlin, Heidelberg.](https://doi.org/10.1007/978-3-642-27851-8_167-1)
> * [Sampling, Aliasing, & Mipmaps - CSCI-6962 Advanced Computer Graphics Cutler](https://www.cs.rpi.edu/~cutler/classes/advancedgraphics/F05/lectures/20_sampling_aliasing.pdf)

Thanks for this feedback, adding a screenshot of other available nodes to illustrate :
image

It has been named "Sample" because as you supposed, it's inspired from the "Sample Mesh" operator.

We could potentially provide two operators LoadGraphicsBuffer (using an unsigned integer) & SampleGraphicsBuffer (using a normalized float). By the way, the "addressing mode" (clamp/wrap/mirror) is missing in this first implementation (It has been pointed in this [document](https://github.com/Unity-Technologies/Graphics/pull/4109#issuecomment-817562938)).
I'm going to raise the naming concern with the team.

Thanks again !

**Comment by landonth** [comment]
@PaulDemeulenaere Sounds great! Thanks so much for the detailed and thoughtful reply. ✨

File: com.unity.visualeffectgraph/Editor/Expressions/VFXExpressionBuffer.cs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+    #pragma warning disable 0659
+    class VFXExpressionSampleBuffer : VFXExpression
+    {
+        public VFXExpressionSampleBuffer() : this(null, VFXValueType.None, string.Empty, VFXValue<GraphicsBuffer>.Default, VFXValue<uint>.Default, VFXValue<uint>.Default, VFXValue<uint>.Default)
+        {
+        }
+
+        //TODOPAUL : Check code convention
```
Comment (PaulDemeulenaere) [inline com.unity.visualeffectgraph/Editor/Expressions/VFXExpressionBuffer.cs:52]: Clean this TODOPAUL

**Review by VladNeykov — APPROVED** [review]
Looking good, no outstanding issues [test doc](https://docs.google.com/document/d/1lkQj67dt_kI6PpjS78UoPLlw8wbtvCQ5VabFFeG5r5w/edit) which can be looked at after the cutoff.

File: com.unity.visualeffectgraph/Runtime/Types/VFXTypes.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using UnityEngine;
+using UnityEngine.VFX;
+
+namespace UnityEngine.VFX
+{
+    [AttributeUsage(AttributeTargets.Struct, AllowMultiple = false)]
```
Comment (PaulDemeulenaere) [inline com.unity.visualeffectgraph/Runtime/Types/VFXTypes.cs:10]: Missing doc here ⬇️

File: com.unity.visualeffectgraph/Runtime/Types/VFXTypes.cs [hunk 3]
```diff
@@ -7,15 +7,24 @@
...
         [Flags]
         public enum Usage
         {
+            /// <summary>Usable as a port type.</summary>
             Default,
+            /// <summary>Usable as a GraphicsBuffer data type.</summary>
             GraphicsBuffer
         }
+
+        /// <summary>Constructor of VFXType.</summary>
```
Comment (JordanL8) [inline com.unity.visualeffectgraph/Runtime/Types/VFXTypes.cs:26]: Initializes and returns an instance of VFXType.

File: com.unity.visualeffectgraph/Runtime/Types/VFXTypes.cs [hunk 4]
```diff
@@ -7,15 +7,24 @@
 
 namespace UnityEngine.VFX
 {
+    /// <summary>
+    /// Attribute to define a VFXType that will be usable in VFX Graph view.
+    /// </summary>
     [AttributeUsage(AttributeTargets.Struct, AllowMultiple = false)]
     public class VFXTypeAttribute : Attribute
     {
+        /// <summary>Usage of VFXType.</summary>
```
Comment (JordanL8) [inline com.unity.visualeffectgraph/Runtime/Types/VFXTypes.cs:16]: Options that define how the Visual Effect Graph can use a VFXType.

File: com.unity.visualeffectgraph/Runtime/Types/VFXTypes.cs [hunk 5]
```diff
@@ -7,15 +7,24 @@
 
 namespace UnityEngine.VFX
 {
+    /// <summary>
+    /// Attribute to define a VFXType that will be usable in VFX Graph view.
```
Comment (JordanL8) [inline com.unity.visualeffectgraph/Runtime/Types/VFXTypes.cs:11]: Defines a VFXType that you can use in Node Workspace.

File: com.unity.visualeffectgraph/Runtime/Types/VFXTypes.cs [hunk 6]
```diff
@@ -7,15 +7,24 @@
...
         public enum Usage
         {
+            /// <summary>Usable as a port type.</summary>
             Default,
+            /// <summary>Usable as a GraphicsBuffer data type.</summary>
             GraphicsBuffer
         }
+
+        /// <summary>Constructor of VFXType.</summary>
+        /// <param name="usages">Usage of the VFXType.</param>
```
Comment (JordanL8) [inline com.unity.visualeffectgraph/Runtime/Types/VFXTypes.cs:27]: Flags that set how the Visual Effect Graph can use the VFXType.

**Review by JordanL8 — CHANGES_REQUESTED** [review]

**Review by JordanL8 — APPROVED** [review]

**Comment by PaulDemeulenaere** [comment]
The needed change in C++ landed (see https://ono.unity3d.com/unity/unity/changeset/5676fbb6fbea) , launching Yamato on really last trunk : https://ono.unity3d.com/unity/unity/changeset/281d883a3da809b5684d799dea25486c92813302 (master didn't update the trunk head yet) on d9458b468bb11c271d7e3da350e66d46ba9bb317 ⏳

**Review by julienf-unity — APPROVED** [review]

**Pull request merged**